cmd/waiter/waiter.go

```diff
@@ -0,0 +1,106 @@
+// Copyright The Shipwright Contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"strconv"
+	"time"
+)
+
+// Waiter represents the actor that will wait for timeout, using a lock-file to keep it actively
+// waiting. When "done" is issued the lock-file is removed and the waiter ends.
+type Waiter struct {
+	flagValues *settings // command-line flags
+}
+
+// ErrTimeout emitted when timeout is reached.
+var ErrTimeout = errors.New("timeout waiting for condition")
+
+// save writes the lock-file with informed PID.
+func (w *Waiter) save(pid int) error {
+	f, err := os.Create(w.flagValues.lockFile)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	log.Printf("Saving '%d' (PID) on '%s' lock-file", pid, w.flagValues.lockFile)
+	pidStr := strconv.Itoa(pid)
+	if _, err = f.WriteString(pidStr); err != nil {
+		return err
+	}
+	return f.Sync()
+}
+
+// read reads the lock-file, must contain an integer.
+func (w *Waiter) read() (int, error) {
+	if _, err := os.Stat(w.flagValues.lockFile); err != nil {
+		return -1, err
+	}
+	data, err := ioutil.ReadFile(w.flagValues.lockFile)
+	if err != nil {
+		return -1, err
+	}
+	pid, err := strconv.Atoi(string(data))
+	if err != nil {
+		return -1, err
+	}
+	return pid, nil
+}
+
+// retry re-execute the informed function once a second until it returns true.
```

Comment:
We might want to run this more frequent. We potentially wait up to near to a second after the source upload is done (and the lock file removed). That's quite a lot. I guess something between 50 and 100 ms would be something we can target.

Reply:
Okay, fair enough! I've modified this routine to wait for 100ms, please consider.